types/widget.ts

```diff
@@ -0,0 +1,23 @@
+import {
+  Helper,
+  HelperState,
+  SearchResults,
+  InstantSearch,
+  SearchParameters,
+} from './instantsearch';
+
+interface WidgetArgs {
+  helper: Helper;
+  state?: HelperState;
+  results?: SearchResults;
+  instantSearchInstance: InstantSearch;
+}
+
+export interface Widget {
+  init(options: WidgetArgs): void;
+  render(options: WidgetArgs): void;
+  dispose(options: any): SearchParameters;
+  getConfiguration?(previousConfiguration: SearchParameters): SearchParameters;
+}
```

Comment:
why not add the routing functions here?

Reply:
The goal was to write the bare minimum to implement the Query Rules connector. We can add them when we need them.